File: C++/542.cpp
__________________________________________________________________________________________________
sample 164 ms submission
static const auto _____ = []()
{
    ios::sync_with_stdio(false);
    cin.tie(0);
    return 0;
}();

class Solution {
public:
    vector<vector<int>> updateMatrix(vector<vector<int>>& matrix) {
        int m = matrix.size(), n = matrix[0].size();
        vector<vector<int>> res(m, vector<int>(n, INT_MAX - 1));
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                if (matrix[i][j] == 0) res[i][j] = 0;
                else {
                    if (i > 0) res[i][j] = min(res[i][j], res[i - 1][j] + 1);
                    if (j > 0) res[i][j] = min(res[i][j], res[i][j - 1] + 1);
                }
            }
        }
        for (int i = m - 1; i >= 0; --i) {
            for (int j = n - 1; j >= 0; --j) {
                if (res[i][j] != 0 && res[i][j] != 1) {
                    if (i < m - 1) res[i][j] = min(res[i][j], res[i + 1][j] + 1);
                    if (j < n - 1) res[i][j] = min(res[i][j], res[i][j + 1] + 1);
                }
            }
        }
        return res;
    }
};
__________________________________________________________________________________________________
sample 20956 kb submission
class Solution {
public:
    vector<vector<int>> updateMatrix(vector<vector<int>>& matrix) {
        vector<int>dirI={0,1,0,-1};
        vector<int>dirJ={1,0,-1,0};
        vector<vector<int>>ans=matrix;
        int m=matrix.size();
        int n=matrix[0].size();
        for(int i=0;i<m;++i)
            for(int j=0;j<n;++j)
                {
                    if(matrix[i][j]!=0)
                    {
                        bool flag=false;
                        for(int k=0;k<4;++k)
                        {
                            int i1=i+dirI[k];
                            int j1=j+dirJ[k];
                            if(i1==-1||i1==m||j1==-1||j1==n)
                                continue;
                            if(matrix[i1][j1]==0)
                            {   flag=true;
                                ans[i][j]=1;
                            }
                        }
                        if(flag)continue;
                        ans[i][j]=10000;
                    }
                }
        for(int i=0;i<m;++i)
            for(int j=0;j<n;++j)
            {
                if(ans[i][j]>=1)
                {
                    for(int k=0;k<4;++k)
                    {
                        int i1=i+dirI[k];
                        int j1=j+dirJ[k];
                        if(i1==-1||i1==m||j1==-1||j1==n)
                                continue;
                        ans[i][j]=min(ans[i][j],ans[i1][j1]+1);
                    }
                }
            }
        for(int i=m-1;i>=0;--i)
            for(int j=n-1;j>=0;--j)
            {
                if(ans[i][j]>=1)
                {
                    for(int k=0;k<4;++k)
                    {
                        int i1=i+dirI[k];
                        int j1=j+dirJ[k];
                        if(i1==-1||i1==m||j1==-1||j1==n)
                                continue;
                        ans[i][j]=min(ans[i][j],ans[i1][j1]+1);
                    }
                }
            }
        return ans;
    }   
};
__________________________________________________________________________________________________

File: Java/542.java
__________________________________________________________________________________________________
sample 4 ms submission
class Solution {
    public int[][] updateMatrix(int[][] matrix) {
        if(matrix == null || matrix.length == 0)
            return matrix;
        int rows = matrix.length;
        int cols = matrix[0].length;
        int [][] output = new int [rows][cols];
        for(int row = 0; row < rows; row++){
            for(int col =0; col < cols; col++){
                if(matrix[row][col] == 1){
                    output[row][col] =  dfs(matrix,row,col,output);
                }
            }
        }
        return output;
    }
    private int dfs(int [][] matrix , int row, int col,int [][] output){
        if(row < 0 || col < 0 || row >= matrix.length || col >= matrix[0].length)
            return Integer.MAX_VALUE;
        if(matrix[row][col] == 0)
            return 1;

        
        // Case 1 Find 0 on top
         if(row > 0 && matrix[row-1][col] == 0) return 1;
         // Case 2 Find 0 on right
         if(col < matrix[0].length-1 && matrix[row][col+1] == 0) return 1;
         // Case 3 Find 0 on bottom
         if(row < matrix.length-1 && matrix[row+1][col] == 0) return 1;
        // Case 4 Find 0 on left
          if(col>0 && matrix[row][col-1] == 0) return 1;
              
        int left,bottom,right,top;
        left=top=9999;
        // if recursive call
        if(row > 0 && output[row-1][col] != 0){
            top = output[row-1][col];
        }
        if(col> 0 && output[row][col-1] != 0){
            left = output[row][col-1];
        }
        bottom = dfs(matrix,row+1, col,output);
        right = dfs(matrix,row,col+1,output);
        int finalMin = Math.min(Math.min(left, right), Math.min(top,bottom))+1;
        return finalMin;
    }
}
__________________________________________________________________________________________________
sample 43016 kb submission
class Solution {
    public int[][] updateMatrix(int[][] matrix) {
        if (matrix== null ||matrix.length == 0)
            return matrix;
        int m = matrix.length;
        int n = matrix[0].length;
        PriorityQueue<int[]> pq = new PriorityQueue<>((a,b) -> Integer.compare(a[2],b[2]));
        int[][] res = new int[m][n];
        for(int i =0; i <m;i++){
            for(int j =0; j <n;j++){
                if (matrix[i][j] == 1){
                    res[i][j] = Integer.MAX_VALUE;
                }else{
                    pq.offer(new int[]{i,j,0});
                }
            }
        }
        
        int[][] dir = {{-1,0},{1,0},{0,1},{0,-1}};
        while(!pq.isEmpty()){
            int[] top = pq.poll();
            int r = top[0];
            int c = top[1];
            int v = top[2];
            for(int[] arr:dir){
                int nr = r+arr[0];
                int nc = c+arr[1];
                if (nr < 0 || nr >= matrix.length || nc < 0 || nc >= matrix[0].length)
                    continue;
                if (v +1 < res[nr][nc]){
                    res[nr][nc] = 1+v;
                    pq.offer(new int[]{nr,nc,1+v});
                }
            }
        }
        return res;
    }
}
__________________________________________________________________________________________________
